core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -906,6 +903,67 @@ void rename(RelNode rel) {
     }
   }
 
+  /**
+   * Checks whether a relexp has made any subset cheaper, and if it so,
+   * propagate new cost to parent rel nodes.
+   *
+   * @param rel       Relational expression whose cost has improved
+   */
+  void propagateCostImprovements(RelNode rel) {
+    RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+    Map<RelNode, RelOptCost> propagateRels = new HashMap<>();
+    PriorityQueue<RelNode> propagateHeap = new PriorityQueue<>((o1, o2) -> {
+      RelOptCost c1 = propagateRels.get(o1);
```

Comment:
After this change, the propagation is neither `BFS` nor `DFS`, can this cause problem such as StackOverFlow when the rel
node hierarchy is very deep ?

Reply:
Although not exactly BFS, but it (Dijkstra) works very similar to BFS, I'd like to view it as a controlled special type of BFS. I think the heap size here and the queue size in BFS should be about the same.

Reply:
It is different with `Dijkstra`, especially when you put different hierarchy nodes into one queue and only sort them with cost.

Reply:
Or, can we make the promotion evidence more clear ? Is there any possibility that we do some benchmark test ?

Reply:
If you consider cost as a kind of distance between relnodes/subsets, this propagation process is basically Dijkstra in a directed graph. Computing the best plan in this directed graph is finding the "shortest" path.

Reply:
I did compare the running time using some big queries, with the patch the whole volcano phase is about 5% faster.

Reply:
Thanks, +1 for this change.

Reply:
Thanks :D